### ISSUE TITLE
networkd-dhcp4: no IPv4 default route with Anonymize=yes

### ISSUE THREAD

**Issue by Waldteufel** [systemd/systemd#7048]
### Submission type

 - Bug report

### systemd version the issue has been seen with

> 235

### Used distribution

> Arch Linux

### In case of bug report: Expected behaviour you didn't see

> systemd-networkd should add a default IPv4 route via the gateway, even when Anonymize=yes is set

### In case of bug report: Unexpected behaviour you saw

> No IPv4 default route is added (same behavior as with UseRoutes=no). As far as I understand, this is due to https://github.com/systemd/systemd/blob/v235/src/network/networkd-network.c#L118 in combination with https://github.com/systemd/systemd/blob/v235/src/network/networkd-dhcp4.c#L83.
> 
> While this behavior is sort-of documented, it makes the anonymization option impractical for IPv4.

### In case of bug report: Steps to reproduce the problem

> set Anonymize=yes in a network file where DHCP is used for IPv4 configuration

**Comment by flokli** [comment]
RFC7844 requests to only send a minimal number of DHCP options. DHCP option 33, which might be used to configure static routes on top of a gateway address is one of those who can be omitted.

This currently gets done by simply disabling `network->dhcp_use_routes`. This does indeed preventing above DHCP option to be requested (not only used, as the name would imply), but also breaks default route insertion later, also see comment [here](https://github.com/systemd/systemd/blob/master/src/network/networkd-dhcp4.c#L637).

/cc @juga0

**Comment by poettering** [comment]
@flokli does it really make sense to suppress option 33? I mean, suppressing that just means that routing won't work properly. I mean, minimizing what we request may be a noble goal, but I am pretty sure we should not do that at the cost of breaking the most basic things.

IIRC dhcp servers aren't supposed to send back records that haven't been requested. that means if you we don't request option 33, we won't get it back either, hence the only way to make sure routing works is to request it...

**Comment by flokli** [comment]
[RFC2132](https://tools.ietf.org/rfc/rfc2132.txt) describes Option 55 (Parameter request list) like this:

> This option is used by a DHCP client to request values for specified
> configuration parameters.  The list of requested parameters is
> specified as n octets, where each octet is a valid DHCP option code
> as defined in this document.

> The client MAY list the options in order of preference.  The DHCP
> server is not required to return the options in the requested order,
> but MUST try to insert the requested options in the order requested
> by the client.

Which doesn't read like the client specifies on what will only accept, more like what things he might especially be interested in, like DNS or NTP Servers. In practice, I often see much more options being pushed than what was requested, and a lot of [those options](http://www.networksorcery.com/enp/protocol/bootp/options.htm) might make sense for the DHCP server to set if deemed useful and for the client to honor if received, even though not explicitly requested (like MTU and TTL options).

Additionally, option 33 is only needed in case you want to add additional static routes on top of the classic Subnet/Gateway scenario. Configuring a gateway normally also works without option 33 being set, but might be required in an ISP-customer scenario with a single /32 lease, and pushing a /32 route to the (on-link) gateway via option 33 (to not waste a /31 per customer).

TLDR: In `Anonymize=yes` mode, networkd should request the options according to (Windows 10's) RFC7844 (implementation), but still honor other options being sent (including option 33).

**Comment by juga0** [comment]
i see two different cases here:

1. to have the option to do not send 33 when ``Anonymize=yes``
2. option 33 is sent but not used when ``Anonymize=yes``

Re. 1., it's commented that some options will not be used when ``Anonymize=yes`` [0] and the PRL will be exactly this [1] in that case.
IMHO it would not make sense to use ``Anonymize=yes`` and then add/remove options in the PRL, as it would make clients to look different and therefore traceable.

Re. 2, it looks like a bug. 
@poettering suggested these 2 PRs [2], [3] to make use of ``UseTimezone`` and ``UseNTP`` when actually they are set (before it was "Always acquire the timezone and NTP").
The problem with ``UseRoutes`` is that it would duplicate option 33 if ``Anonymize=yes`` and ``UseRoutes=yes``. However, i think i forgot to take into account the cases where ``UseRoutes=yes`` or ``Anonymize=yes`` here [4], which i think can be solved adding ``link->network->dhcp_use_routes || link->network->anonymize``

[0] https://github.com/systemd/systemd/blob/master/man/systemd.network.xml#L1070, https://github.com/systemd/systemd/blob/master/src/network/networkd-network.c#L212
[1] https://github.com/systemd/systemd/blob/master/src/libsystemd-network/sd-dhcp-client.c#L120
[2] https://github.com/systemd/systemd/pull/6725
[3] https://github.com/systemd/systemd/pull/6726
[4] https://github.com/systemd/systemd/blob/master/src/network/networkd-dhcp4.c#L83, https://github.com/systemd/systemd/blob/master/src/network/networkd-dhcp4.c#L193

**Comment by Waldteufel** [comment]
I would like to point out that my original problem is related to **option 3** (SD_DHCP_OPTION_ROUTER), not option 33 (SD_DHCP_OPTION_STATIC_ROUTE).

@juga0 regarding your idea: As [sd_dhcp_client_set_request_option](https://github.com/systemd/systemd/blob/4c701096002fff540d9ddb3b21398c551ac3af78/src/libsystemd-network/sd-dhcp-client.c#L200) fails when the option is already present, wouldn't handling EEXIST separately in [networkd-dhcp4.c:643](https://github.com/systemd/systemd/blob/fc5ffacdabbcbbb18f2f249512d8c5a30ed6bc66/src/network/networkd-dhcp4.c#L643) be a cleaner way to handle this? This way you could avoid the error when adding a duplicate option, while keeping the original meaning of `dhcp_use_routes`.

This could also help to keep the code readable. At the moment, I find conditions like in [networkd-dhcp4.c:640](https://github.com/systemd/systemd/blob/fc5ffacdabbcbbb18f2f249512d8c5a30ed6bc66/src/network/networkd-dhcp4.c#L640) a bit confusing, as [network_apply_anonymize_if_set](https://github.com/systemd/systemd/blob/fc5ffacdabbcbbb18f2f249512d8c5a30ed6bc66/src/network/networkd-network.c#L118) already sets `dhcp_use_routes = false` when `dhcp_anonymize` is set.

**Comment by flokli** [comment]
There is an `int anonymize` at `sd_dhcp_client_new` that does apply an initial set of options in default and anonymize case. Can't we just return from `dhcp4_configure` after the `sd_dhcp_client_set_ifindex` in anonymize case and skip setting more options alltogether, as this would allow fingerprinting?

In addition, `network_apply_anonymize_if_set` sets `network->dhcp_use_routes` to `false`, although we do send DHCP options 3, 33 and 121. This looks wrong.

**Comment by juga0** [comment]
@Waldteufel: In this branch [1] i've changed what i think you meant, it makes sense to me, would you mind to try it and comment whether now the routes are used properly?.
Thanks for catching this bug.
[1] https://github.com/systemd/systemd/compare/master...juga0:fixes/7048?expand=1

**Comment by juga0** [comment]
@flokli i appreciate your comments but i think that deciding which options to include or not is a different topic not really related to this bug.

**Comment by Waldteufel** [comment]
@juga0: I see no explicit handling of -EEXIST in your code, so dhcp4_configure would abort (`return r` in [Line 644](https://github.com/systemd/systemd/blob/fc5ffacdabbcbbb18f2f249512d8c5a30ed6bc66/src/network/networkd-dhcp4.c#L644)) after trying to add the option. I will have a closer look over the weekend.

**Comment by juga0** [comment]
@Waldteufel You're right, i forgot to check EEXITS, though instead of checking what r returns, i'd live it like previously, just check whether anonymize is true before adding those options.
In this case the only thing to modify is to ensure that dhcp_use_routes=1 when using Anonymize. I added the comment in the wrong function, BTW.
You can see what i mean in this branch: https://github.com/systemd/systemd/compare/master...juga0:fixes/7048-2?expand=1

**Comment by Waldteufel** [comment]
Ok, I can confirm that enforcing dhcp_use_routes=1 solves my problem.
